### PR TITLE
telemetry: add post process for metric

### PR DESCRIPTION
**Pull request by zirain**
**Please provide a description of this PR:**

fixes: https://github.com/istio/istio/issues/46858

### CODE REVIEW

File: pilot/pkg/model/telemetry.go [hunk 1]
```diff
@@ -855,9 +857,94 @@ func mergeMetrics(metrics []*tpb.Metrics, mesh *meshconfig.MeshConfig) map[strin
...
+	for k := range tpb.MetricSelector_IstioMetric_name {
+		if k == int32(tpb.MetricSelector_ALL_METRICS) {
+			continue
+		}
+		res |= 1 << int(k-1)
+	}
+	return res
+}()
+
+func simplyMetricConfig(cfg metricConfig) metricConfig {
```
Comment (howardjohn) [inline pilot/pkg/model/telemetry.go:887]: This algorithm is not very obvious what it does. Can we add comments about what we are doing?

Comment (zirain) [reply pilot/pkg/model/telemetry.go:890]: updated

**Comment by kyessenov** [comment]
I started looking but it's complex logic and I got swamped by all the PRs, so please wait.

**Comment by zirain** [comment]
> I started looking but it's complex logic and I got swamped by all the PRs, so please wait.

@howardjohn @kyessenov do you have bandwith for this? want to land this in v1.23.

**Comment by keithmattix** [comment]
I will try and look as well

File: pilot/pkg/model/telemetry.go [hunk 2]
```diff
@@ -855,9 +857,97 @@ func mergeMetrics(metrics []*tpb.Metrics, mesh *meshconfig.MeshConfig) map[strin
...
+	var processed []metricsOverride
+	tagKeys := make(map[tagOverride]int)
+	for _, override := range cfg.Overrides {
+		if override.Disabled || len(override.Tags) == 0 {
+			continue
+		}
+		processed = append(processed, override)
+
+		idx := slices.Index(allMetrics, override.Name)
+		for _, tag := range override.Tags {
```
Comment (keithmattix) [inline pilot/pkg/model/telemetry.go:900]: This next bit of code is the most confusing part of this algorithm; please add more comments

File: pilot/pkg/model/telemetry.go [hunk 3]
```diff
@@ -814,7 +816,7 @@ func mergeMetrics(metrics []*tpb.Metrics, mesh *meshconfig.MeshConfig) map[strin
 			}
 
 			for metric, override := range metricMap {
-				tags := []tagOverride{}
+				tags := make([]tagOverride, 0, len(metricToPrometheusMetric))
```
[the remark this answers is not in the document]
Comment (zirain) [reply pilot/pkg/model/telemetry.go:819]: in most of the case it won't, and it's fine for a slice to exceed?

File: pilot/pkg/model/telemetry.go [hunk 4]
```diff
@@ -855,9 +857,97 @@ func mergeMetrics(metrics []*tpb.Metrics, mesh *meshconfig.MeshConfig) map[strin
 		})
 		processed[provider] = tmm
 	}
+
+	processed = postProcessMetrics(processed)
 	return processed
 }
 
+// postProcessMetrics tries to minimize the metricsConfig
```
[the remark this answers is not in the document]
Comment (zirain) [reply pilot/pkg/model/telemetry.go:865]: already there under function `simplyMetricConfig`

File: pilot/pkg/model/telemetry.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (keithmattix) [reply pilot/pkg/model/telemetry.go:819]: Ha, you're right; I've been writing too much C++ 🙂

File: pilot/pkg/model/telemetry.go (hunk 2 above)
Comment (keithmattix) [reply pilot/pkg/model/telemetry.go:904]: Ah the comments are helpful. I'm still struggling a bit to grasp all of the logic. Maybe @kyessenov or @howardjohn have some ideas for refactoring to make it clearer?

**Comment by kyessenov** [comment]
I'm not super familiar with telemetry API control plane logic. Can we ask maintainers to approve it? @keithmattix @lei-tang ?

File: pilot/pkg/model/telemetry.go [hunk 5]
```diff
@@ -855,9 +857,103 @@ func mergeMetrics(metrics []*tpb.Metrics, mesh *meshconfig.MeshConfig) map[strin
...
+
+// maxVal equals to 2^(len(tpb.MetricSelector_IstioMetric_name)-1) - 1, currently it is 1023
+var maxVal = func() int {
+	res := 0
+	for k := range tpb.MetricSelector_IstioMetric_name {
+		// skip tpb.MetricSelector_ALL_METRICS, it's not included in allMetrics
+		if k == int32(tpb.MetricSelector_ALL_METRICS) {
+			continue
+		}
+		res |= 1 << int(k-1)
```
[the remark this answers is not in the document]
Comment (zirain) [reply pilot/pkg/model/telemetry.go:885]: let me try when I'm free.

**Comment by zirain** [comment]
not stale

**Pull request closed without merging**